blockchain/process.go

```diff
@@ -70,6 +71,53 @@ func (b *BlockChain) maybeUpdateAssumeValid(node *blockNode) {
 		node.height)
 }
 
+// isAssumeValidAncestor determines whether the provided node is both an
+// ancestor of the assume valid node and an ancestor of the best header.  False
+// is returned when the assume valid node has not been discovered or when assume
+// valid is disabled.
+//
+// Note: The assume valid node is clamped back as needed to ensure that it is at
+// least 2 weeks worth of blocks behind the best header.  This protects against
+// malicious actors trying to play games by tricking users into using an assume
+// valid hash near the tip.
+//
+// This function MUST be called with the chain lock held (for reads).
+func (b *BlockChain) isAssumeValidAncestor(node *blockNode) bool {
+	// Return false immediately if the assume valid node is not set.
+	if b.assumeValidNode == nil {
+		return false
+	}
+
+	// Return false if the node is not an ancestor of the best header.
+	b.index.RLock()
+	bestHeader := b.index.bestHeader
+	b.index.RUnlock()
+	if !node.IsAncestorOf(bestHeader) {
+		return false
+	}
+
+	// Calculate the expected number of blocks in 2 weeks.
+	const timeInTwoWeeks = time.Hour * 24 * 14
+	targetTimePerBlock := b.chainParams.TargetTimePerBlock
+	expectedBlocksInTwoWeeks := int64(timeInTwoWeeks / targetTimePerBlock)
+
+	// Return false if the best header height is less than the number of blocks
+	// expected in 2 weeks.
+	if bestHeader.height < expectedBlocksInTwoWeeks {
+		return false
+	}
+
+	// Clamp the assume valid node back as needed to ensure that it is at least
+	// 2 weeks worth of blocks behind the best header.
+	assumeValidNode := b.assumeValidNode
+	clampToHeight := bestHeader.height - expectedBlocksInTwoWeeks
+	if assumeValidNode.height > clampToHeight {
+		assumeValidNode = bestHeader.Ancestor(clampToHeight)
```

Comment:
This doesn't look right to me.  What if `assumeValidNode` is not already an ancestor of the best header?  It might have been found on a side chain node and this would end up choosing the wrong one.

e.g.
```
a -> b  -> c  -> ... -> best header (height: 4034)
 \-> b' -> c' -> d' (assumeValidNode, height: 3)
```

Now, assume `expectedBlocksInTwoWeeks = 4032`, then `clampToHeight = 4034-4032 = 2` and `bestHeader.Ancestor(2) = c`, not `c'` as I would expect.  That would result in `b` incorrectly being assumed as valid.

Thus, I believe this should be `assumeValidNode = assumeValidNode.Ancestor(clampToHeight)`.

Reply:
You are correct, good catch.  The intention was to clamp it back from the assumed valid node and not the best header, as you pointed out.  Fixed.